### PR TITLE
Classes documented using KDoc

### PR DESCRIPTION
**Pull request by FrankieBDev**

### CODE REVIEW

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 1]
```diff
@@ -0,0 +1,349 @@
+/**
+ * Enum class with available car colours
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: **Minor**: I'd remove "enum class" clarification, as it feels quite obvious when you are using it and doesn't contribute much to the doc itself. One sentence explaining "**what**, **why** and **where**"(_order of answering those questions might differ of course_) should be enough. Something like "Colours available as part of car representation/specification." Or simply leaving it as is just without mentioning that it's an enum should work as well. 

Even though Car and AvailableColours are in the same file, could you please leave a link to Cars class from the doc, so that user-developer can easily navigate there?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+ */
+enum class AvailableColours { Red, Blue, Green, Yellow, Silver, White, Black }
+
+/**
+ * Superclass: Car
+ * @property year year of manufacture
+ * @property colour exterior colour
+ * @property price base price
+ */
+// Superclass
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: I'd elaborate a bit more on the explanation what is a car. Mentioning that's a superclass is alright, but twice is not :) 
Try to follow the same rule I mentioned above. Once again, one sentence should be enough.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+ */
+abstract class BMW(val fuelType: String, val gearBox: String) : Car(1999, AvailableColours.Blue, 400.00) {
+    /**
+     * @property price price of the car based on fuel type and gearbox properties.
+     *  If car runs on petrol AND is automatic base price is multiplied by 30.
+     *  If car has alternative properties price is base price plus 450.
+     *
+     */
+    override var price: Double = 0.0
+        get() = if (fuelType == "Petrol" && gearBox == "Automatic") {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Could you move some of the hardcoded strings into constants? (as they are repeated in several places)

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+/**
+ * Subclass Mercedes Manufacturer type car.  Extends Car class.
+ * @property doors number of doors
+ * @property owners number of previous owners
+ */
+open class Mercedes(val doors: Int, val owners: Int) : Car(2010, AvailableColours.Green, 500.00) {
+    /**
+     * @property motCheck checks for valid MOT
+     */
+    val motCheck: Boolean = true
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: So MOT check will be just always true?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+     * Abstract function to navigate the vehicle.
+     */
+    abstract fun navigate()
+}
+/**
+ * Subclass Mercedes Manufacturer type car.  Extends Car class.
+ * @property doors number of doors
+ * @property owners number of previous owners
+ */
+open class Mercedes(val doors: Int, val owners: Int) : Car(2010, AvailableColours.Green, 500.00) {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: So it won't be possible to have Mercedes with any other colour? Only green once?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    /**
+     * @property motCheck checks for valid MOT
+     */
+    val motCheck: Boolean = true
+
+    /**
+     * @property doorPremium multiplies number of doors by 1.5
+     */
+    val doorPremium: Double
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Could you please explain for what purpose we need this property? More as if in real life example. And why do we need to have that as getter/

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        get() = doors * 1.5  // Getter
+    /**
+     * @property price price of the car based on doors property.
+     *  If a car has less than 5 doors 10 is added to base price.
+     *  If car has 5 to 8 doors 20 is added to base price, if there are more than 8 doors 33 is added.
+     */
+    override var price: Double = 0.00 + if (doors >= 5) {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Why do we need the `0.0 +` before `if`?
Is there a way to tidy this up? Maybe using a different construction or somehow lower the amount of if-else clauses.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+                super.price + 33.00
+            } else super.price + 20.00
+        } else {
+            super.price + 10.00
+        }
+}
+/**
+ * Subclass Ford Manufacturer type car.  Extends Car class.
+ */
+open class Ford : Car(2020, AvailableColours.Red, 300.00) {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: All Ford cars will be 2020 year, feels like limitation to me.
Can we have the year definition not depending on the make or even model?
I'd say year is something that should be defined during specific car(read object) manufacturing(read creation).

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    /**
+     * @property seats number of seats
+     * @property location pickup location
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: `seats` prop has also the documentation for `location`, which shouldn't be the case.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+     */
+    var seats: Int = 8
+    val location: String = "Manchester"
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: `location` is `val`, which in turn means it's there forever and as a consequence all `Ford` cars will be located in Manchester. Can we make it more flexible?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+     * Navigation function which prints a message dependent on whether the dashboard is big or small.
+     */
+    override fun navigate() {
+        if (insurance) {
+            if (dashboard == "big") {
+                price * 40
+                println("Beta is so easy to navigate")
+            } else {
+                price / 3
+                println("Beta map is too small")
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Those price operations don't result in anything. Price isn't reassigned, that's literally an "empty" execution, nothing happens.
On a side note - even if the the result of execution would have been assigned to the `price` prop, why would we need to change the price while navigating?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+ * @property dashboard Big or small dashboard type
+ * @property ecoRating Eco Rating Score
+ */
+class Omega(private val windowSize: Int, private val dashboard: String, val ecoRating: Int) : BMW("Petrol", "Manual") {
+    /**
+     * Navigation function which prints a message dependent on whether the dashboard is big or small.
+     */
+    override fun navigate() {
+        if (dashboard == "big") {
+            price * 20
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Same as above.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+            price / 13
+            println("Omega map is too small")
+        }
+    }
+}
+
+/**
+ * Subclass Mayback Model type car.  Extends Mercedes class.
+ */
+class Mayback : Mercedes(4, 3) {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: **Minor**: I think that's Maybach 😄 
4 and 3 is a bit hard to read(kinda [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))), especially during review. Usually it's helpfull to specify the param explicitly, by using  [named arguments](https://kotlinlang.org/docs/functions.html#named-arguments).  
In this case you would end up having something like:
`Mercedes(doors = 2, owners = 2)`
Initially, named arguments where introduced for a bit diff reasons, but using them this way will increase readability.
Bear in mind IDEA is smart enough to highlight those params even without specifying them directly, but that's not the case for "raw" text.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+     */
+    fun sprinterMaxSpeedPrice(): Double {
+        val extras: Double = when (maxSpeed) {
+            in 0..60 -> 10.00
+            in 61..80 -> 20.00
+            in 81..100 -> 30.00
+            else -> 50.00
+        }
+        return extras
+    }
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Why do we need the `extras` prop? Can't we `return` the when block directly?
Do you need to have those two methods publicly available? It seems to be an internal implementation detail.

Also both fun naming isn't obvious, could you rephrase it a bit to describe what exactly is happening. It should be exhaustive(like your doc). And in most cases if the naming is exhaustive you don't need to document it.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+ *  @property rating Customer rating score.
+ *  @property maxSpeed Maximum speed of vehicle.
+ */
+class Traveliner(private val rating: Int, val maxSpeed: Int) : Mercedes(6, 1) {
+    /**
+     * @param rating Customer rating score.
+     * @param maxSpeed Maximum speed of vehicle.
+     * @param gearBox Gearbox type.
+     */
+    constructor(rating: Int, maxSpeed: Int, gearBox: String) : this(rating, maxSpeed) {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: `gearBox` is never used, it's just a param which feels redundant here.
Also the `Traveliner` and `Sprinter` secondary constructors are the same, so it's worth re-using that part somehow or adjusting specific class that differs. 
**General advice** - if you are copying something(not necessary here) it's worth thinking about re-usability straight away to cut off any possible duplication, which is basically increasing the chance of introducing a bug/defect at least twice.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+        println(specs)
+    }
+    /**
+     * @property price Base price plus max speed price calculation.
+     */
+    override var price: Double = super.price + travelinerMaxSpeedPrice()
+    /**
+     * Calculates a price based on max speed.
+     */
+    fun travelinerMaxSpeedPrice(): Double {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Same here, shouldn't it be internal implementation?

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+class Cougar(private val interior: String, private var materialPrice: Double, val spareTyre: Boolean, val radioType: String) :
+    Ford() {
+    /**
+     * @property price Base price plus interior calculation
+     */
+    override var price: Double = super.price + interiorPremium()
+    /**
+     * Calculates a price based on interior.  If the interior is Leather or Luxuary the price is multiplied by 2.5.
+     */
+    fun interiorPremium(): Double {
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: Same as above. Feel free to resolve if you plan on exposing those kind of functions.
But there should be a strong explanation on why we would need to do that.

File: app/src/main/java/com/fbada/projectone/Car.kt [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+    /**
+     * @property price Base price plus interior price including VAT.
+     */
+    override var price: Double = super.price + interiorPriceAddVAT()
+    /**
+     * Calculates added VAT into interior price.
+     */
+    fun interiorPriceAddVAT(): Int {
+        val vat = interiorPrice / 100 * 20
+        val priceWVAT = interiorPrice + vat
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/Car.kt]: `priceWVAT` doesn't fall under kotlin prop declaration [convention](https://kotlinlang.org/docs/coding-conventions.html).
And TBH it feels like you can simply `return interiorPrice + vat`
The function itself is doing two things adding the VAT to the price and returning it, however in the declaration we state only one thing. In addition we aren't really affecting any price outside of this function, which means the scope and affect of this change is really limited, thus if that is planned to be an exposed fun it's a bit missleading.
In any case I'd probably go with something like `interiorPriceWithVAT` or `interiorPriceIncludingVAT`.

**Review by slowpacer — COMMENTED** [review]
Lots of properties and functions aren't used, obviously because they aren't fully integrated yet, but please make sure you aren't exposing something that should be hidden. In other words double check your visibility identifiers.

File: app/src/main/java/com/fbada/projectone/domain/Car.kt [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+enum class AvailableColours { Red, Blue, Green, Yellow, Silver, White, Black }
+
+const val petrol = "Petrol"
+const val auto = "Automatic"
+const val manual = "Manual"
+const val big = "Big"
+const val leather = "Leather"
+const val lux = "luxury"
+/**
+ * Superclass: [Car]
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/domain/Car.kt]: Agreed to remove car link.

File: app/src/main/java/com/fbada/projectone/domain/Car.kt [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+ */
+enum class AvailableColours { Red, Blue, Green, Yellow, Silver, White, Black }
+
+const val petrol = "Petrol"
+const val auto = "Automatic"
+const val manual = "Manual"
+const val big = "Big"
+const val leather = "Leather"
+const val lux = "luxury"
+/**
```
Comment (slowpacer) [inline app/src/main/java/com/fbada/projectone/domain/Car.kt]: Should be uppercase and placed at the very top of the file.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 4 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Changed to var

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 5 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Updated all subclasses so that specific colours and years are no longer hardcoded and instead now inherit the properties.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 6 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Property is unnecessary.  Deleted.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 7 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: refactored and used when block instead.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 8 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: edited with colours

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 10 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Changed to var

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 11 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Code block doesn't make logical sense.  Refactored to make more sense.  May further refactor with hierarchy refactor.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 12 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Same comment as above

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 13 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Whoops- changed to the actual model, rather than the one I made up! XD

Changed all changed all constructors to include named arguments.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 14 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: All suggested improvements completed.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 15 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Adjusted so gearbox is used.

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 16 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: values edited for differentiation

File: app/src/main/java/com/fbada/projectone/Car.kt (hunk 18 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/Car.kt]: Changed naming and refactored code as suggested.

File: app/src/main/java/com/fbada/projectone/domain/Car.kt (hunk 20 above)
Comment (FrankieBDev) [reply app/src/main/java/com/fbada/projectone/domain/Car.kt]: renamed
